foundationdb/src/database.rs

```diff
@@ -49,6 +51,45 @@ impl Database {
             Ok(Transaction::new(self.clone(), trx))
         }
     }
+
+    /// `transact` returns a future which retries on error. It tries to resolve a future created by
+    /// caller-provided function `f` inside a retry loop, providing it with a newly created
+    /// transaction. After caller-provided future resolves, the transaction will be committed
+    /// automatically.
+    pub fn transact<F, Fut, Item>(&self, f: F) -> Box<Future<Item = Fut::Item, Error = FdbError>>
+    where
+        F: FnMut(Transaction) -> Fut + 'static,
+        Fut: Future<Item = Item, Error = FdbError> + 'static,
+        Item: 'static,
+    {
+        let db = self.clone();
+
+        let f = result(db.create_trx()).and_then(|trx| {
+            loop_fn((trx, f), |(trx, mut f)| {
+                let trx0 = trx.clone();
+                f(trx.clone())
+                    .and_then(move |res| {
+                        // try to commit the transaction
+                        trx0.commit().map(|_| res)
+                    })
+                    .then(|res| match res {
+                        Ok(v) => {
+                            // committed
+                            Ok(Loop::Break(v))
+                        }
+                        Err(e) => {
+                            if e.should_retry() {
```

Comment:
Should this loop be bounded? it looks like currently it will potentially be an infinite loop in the worst case...

Reply:
You're right. As an example, if we remove `RetryLimit` from `test_transact` it will run indefinitely. I believe it's okay to leave it as is because the behavior matches with other implementations[1].

[1] https://github.com/apple/foundationdb/blob/master/bindings/go/src/fdb/database.go#L86-L104

Reply:
Can we add a comment documenting the reliance on the RetryLimit as the upper bound for this? That way people won't accidentally be caught off-guard by it.

Reply:
Done.